### PR TITLE
Adds docs for TypeGuard type, #10591

### LINKED ISSUE #10591: Document TypeGuard (PEP 647)
We don't have documentation about TypeGuard ([PEP 647](https://www.python.org/dev/peps/pep-0647/)). It will be included in Python 3.10, and it would be good to have at least basic documentation after Python 3.10 is out, at the latest.

### PR DESCRIPTION
**Pull request by sobolevn**
I am writting an article about ``TypeGuard``s and I've noticed that `mypy` does not document how `TypeGuard`s work.

So, here it is!

It has several examples from the PEP itself and several custom ones.
I have tried to cover all corner-cases that I am aware of, but possibly there are other details that I've missed.

I would appreciate your feedback! 👍 

Closes #10591

### CODE REVIEW

File: docs/source/casts.rst [hunk 1]
```diff
@@ -47,3 +47,149 @@ any operations on the result. For example:
...
+
+.. code-block:: python
+
+  from typing import List
+
+  def is_str_list(val: List[object]) -> bool:
+    """Determines whether all objects in the list are strings"""
+    return all(isinstance(x, str) for x in val)
+
+  def func1(val: List[object]):
```
Comment (JelleZijlstra) [inline docs/source/casts.rst:73]:
```suggestion
  def func1(val: List[object]) -> None:
```

File: docs/source/casts.rst [hunk 2]
```diff
@@ -47,3 +47,149 @@ any operations on the result. For example:
...
+.. code-block:: python
+
+  from typing import List
+  from typing import TypeGuard  # use `typing_extensions` for `python<3.10`
+
+  def is_str_list(val: List[object]) -> TypeGuard[List[str]]:
+      """Determines whether all objects in the list are strings"""
+      return all(isinstance(x, str) for x in val)
+
+  def func1(val: List[object]):
```
Comment (JelleZijlstra) [inline docs/source/casts.rst:89]:
```suggestion
  def func1(val: List[object]) -> None:
```

File: docs/source/casts.rst [hunk 3]
```diff
@@ -47,3 +47,149 @@ any operations on the result. For example:
...
+      return all(isinstance(x, type) for x in val)
+
+  items: Set[Any]
+  if is_set_of(items, str):
+      reveal_type(items)  # Set[str]
+
+TypeGuards as methods
+---------------------
+
+Method can also serve as the ``TypeGuard``:
```
Comment (JelleZijlstra) [inline docs/source/casts.rst:162]:
```suggestion
 A method can also serve as the ``TypeGuard``:
```

File: docs/source/casts.rst [hunk 4]
```diff
@@ -47,3 +47,149 @@ any operations on the result. For example:
...
+
+.. code-block:: python
+
+  class StrValidator:
+      def is_valid(self, instance: object) -> TypeGuard[str]:
+          return isinstance(instance, str)
+
+  def func(to_validate: object):
```
Comment (JelleZijlstra) [inline docs/source/casts.rst:170]:
```suggestion
  def func(to_validate: object) -> None:
```

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
Thanks! I have a couple of wording suggestions.

Also, I'm not sure the "Casts" section is the right place for these docs. "More types" is the traditional place for advanced type system features. I'm curious to hear others' opinions.

**Comment by sobolevn** [comment]
@JelleZijlstra thanks! I thought that "casts" is the closest to type guards. I can move it, no problem.

File: docs/source/casts.rst [hunk 5]
```diff
@@ -47,3 +47,149 @@ any operations on the result. For example:
     x.whatever()  # Type check error
     y = cast(Any, x)
     y.whatever()  # Type check OK (runtime error)
+
+
+User-Defined Type Guards
+************************
+
+MyPy supports User-Defined Type Guards
```
Comment (hauntsaninja) [inline docs/source/casts.rst:55]:
```suggestion
Mypy supports User-Defined Type Guards
```

File: docs/source/casts.rst [hunk 6]
```diff
@@ -47,3 +47,149 @@ any operations on the result. For example:
...
+(`PEP-647 <https://www.python.org/dev/peps/pep-0647/>`_).
```
Comment (hauntsaninja) [inline docs/source/casts.rst:56]:
```suggestion
(:pep:`647`).
```

File: docs/source/casts.rst [hunk 7]
```diff
@@ -47,3 +47,149 @@ any operations on the result. For example:
...
+
+What's a type guard?
+It is a way for programs to influence conditional
```
Comment (hauntsaninja) [inline docs/source/casts.rst:59]:
```suggestion
A type guard is a way for programs to influence conditional
```

File: docs/source/casts.rst [hunk 8]
```diff
@@ -47,3 +47,149 @@ any operations on the result. For example:
...
+
+  from typing import List
+
+  def is_str_list(val: List[object]) -> bool:
+    """Determines whether all objects in the list are strings"""
+    return all(isinstance(x, str) for x in val)
+
+  def func1(val: List[object]) -> None:
+      if is_str_list(val):
+          reveal_type(val)  # List[object]
```
Comment (hauntsaninja) [inline docs/source/casts.rst:75]:
```suggestion
          reveal_type(val)  # Reveals List[object]
```

File: docs/source/casts.rst [hunk 9]
```diff
@@ -47,3 +47,149 @@ any operations on the result. For example:
...
+          print(" ".join(val)) # Error: invalid type
```
Comment (hauntsaninja) [inline docs/source/casts.rst:76]:
```suggestion
          print(" ".join(val)) # Error: incompatible type
```

File: docs/source/casts.rst [hunk 10]
```diff
@@ -47,3 +47,149 @@ any operations on the result. For example:
...
+
+The same example with ``TypeGuard``:
+
+.. code-block:: python
+
+  from typing import List
+  from typing import TypeGuard  # use `typing_extensions` for `python<3.10`
```
Comment (hauntsaninja) [inline docs/source/casts.rst:83]:
```suggestion
  from typing import TypeGuard  # use `typing_extensions` for Python 3.9 and below
```

File: docs/source/casts.rst [hunk 11]
```diff
@@ -47,3 +47,149 @@ any operations on the result. For example:
...
+  `not strict <https://www.python.org/dev/peps/pep-0647/#enforcing-strict-narrowing>`_.
+  For example, you can narrow ``str`` to ``int``:
+
+  .. code-block:: python
+
+    def f(value: str) -> TypeGuard[int]:
+        return True
+
+  It was noted that without enforcing strict narrowing,
+  it would be possible to break type safety.
```
Comment (hauntsaninja) [inline docs/source/casts.rst:109]:
```suggestion
  Note: since strict narrowing is not enforced, it's easy
  to break type safety.
```

**Comment by hauntsaninja** [comment]
Thank you for this! Fixed a couple small nits

**Comment by sobolevn** [comment]
@JelleZijlstra @hauntsaninja so, should I move this to "Advanced types"? Or is "Casts" the proper place? 🙂

**Comment by hauntsaninja** [comment]
I'd probably put it in "More types" too, but don't feel strongly.

**Comment by sobolevn** [comment]
Moved! 👍

**Comment by sobolevn** [comment]
@hauntsaninja @JelleZijlstra friendly ping 🙂

**Comment by hauntsaninja** [comment]
Thanks again! :-)

**Pull request merged**